### DISCUSSION TITLE
Roadmap to v4.0 - jQuery removal 🚀

### DISCUSSION THREAD

**Discussion by ghiscoding** [6pac/SlickGrid#743] [Ideas]
## Roadmap to jQuery removal (v4.0)
This is a work in progress, though I did manage to get it all working in [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) and even removed jQuery entirely on that project and it works, so I know it's achievable in SlickGrid as well (here). Thanks to @MarkoBL for starting the work and making the biggest work of migrating all the core files in PR #734 (however that did not include any Controls/Plugins, so these have to be migrated too in separate PRs). 

We probably don't want to release a new version 4.0 without migrating everything to vanilla JS. The goal is to make it happen sometime around June of this year.

**Note** all PRs (Pull Requests) should be done against the **[`next`](https://github.com/6pac/SlickGrid/tree/next)** branch and all Cypress E2E tests must still be passing

#### Core (9/9, all done in PR #734, thanks to @MarkoBL)
- [x] slick.grid.js 
- [x] slick.core.js
- [x] slick.compositeeditor.js
- [x] slick.dataview.js
- [x] slick.interaction.js
- [x] slick.groupitemmetadataprovider.js
- [x] slick.remotemodel.js
   - skipped since this one since it's too specific to jQuery with JSONP
- [x] slick.editors.js
- [x] slick.formatters.js

#### Controls (3/3)
- [x] slick.columnpicker.js (ref PR #752)
- [x] slick.gridmenu.js (ref PR #752)
- [x] slick.pager.js (ref PR #762)

#### Plugins (19/19)
- [x] slick.autotooltips.js (ref PR #745)
- [x] slick.cellcopymanager.js (ref PR #746)
- [x] slick.cellexternalcopymanager.js (ref PR #746)
- [x] slick.cellmenu.js (ref PR #753)
- [x] slick.cellrangedecorator.js (ref PR #754)
- [x] slick.cellrangeselector.js (ref PR #754)
- [x] slick.cellselectionmodel.js (ref PR #754)
- [x] slick.checkboxselectcolumn.js (ref PR #755)
- [x] slick.contextmenu.js (ref PR #753)
- [x] slick.crossgridrowmovemanager.js (ref PR #756)
- [x] slick.customtooltip.js (ref PR #747)
- [x] slick.draggablegrouping.js (ref PR #744)
- [x] slick.headerbuttons.js (ref PR #748)
- [x] slick.headermenu.js (ref PR #748)
- [x] slick.resizer.js (ref PR #758)
- [x] slick.rowdetailview.js (ref PR #760)
- [x] slick.rowmovemanager.js (ref PR #756)
- [x] slick.rowselectionmodel.js (ref PR #754)
- [x] slick.state.js (ref PR #757)

#### Others
- [x] get it all working in [Slickgrid-Universal](https://github.com/ghiscoding/slickgrid-universal) without jQuery
   - all unit tests & Cypress E2E tests are now passing and jQuery was entirely removed
- [x] remove jQuery dependency from `package.json`
- [x] remove jQuery from all possible examples (ref PR #767)
   - [x] remove jQuery/jQueryUI from `/lib` folder and replace with CDN for the examples that still require it (ref PR #771)
   - [x] maybe keep jQuery event drag but add their license into folder
- [x] we should include PR #765 as well
- [x] remove Map polyfill in `slick.core.js` since we will target ES6 and Map is supported (ref PR #759)
- [x] remove an `eval()` function that was added and could be a security risk
   - this came from X-SlickGrid merge, it serves no purpose and it is safe to remove
- [x] look at replacing [.grep()](https://api.jquery.com/jQuery.grep/) with vanilla `.filter()` if possible
- [x] revisit all jQuery `.each` that were replaced by `forEach` since their previous `return` have to be replaced with `continue`
   - I think this one got covered by @6pac, checked if confirmed
- [x] replace all `Slick.Utils.template` with `createDomElement` function (ref PR #766)
- [x] bug found with grid scrolling 
- [x] possibly address this issue #559
- [x] this will also close the issue #388
- [x] add better aria accessibility (ref PR #764)
- [x] add a note in issue #515 that filling the window with grid is a one time thing, the alternative is to use the Grid Resizer and we should add a note in the filling window example
- [x] need to test if having column name with html code still works
- [x] possibly look into the issue #238
- [x] look at using `passive` event listener with scroll, mouse, ... for better perf, see [SO](https://stackoverflow.com/a/37721906/1212166) and [MDN](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#improving_scrolling_performance_with_passive_listeners) details (ref PR #769)
   - also probably better to add check if browser supports `passive` listener, see [MDN](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#safely_detecting_option_support) and [alternative](https://github.com/rafgraph/detect-it/blob/a90c09ae774dab5e39a18ff597401c01bb73bca5/src/index.ts#L22-L35)
   - this will probably close issue #199
- [x] close #393 since we no longer require any 3rd party, apart from SortableJS which is used as CDN and not part of `/lib` folder
- [x] write up a Migration Guide and add link to homepage
   - we no longer provide jQuery (removed from lib folder, use CDN)
   - here's the [Migration Guide v4.0](https://github.com/6pac/SlickGrid/wiki/Major-version-4.0---Removal-of-jQuery-requirement)
- [x] provide another styling theme that is a bit more modern (for example [vue-slimgrid](https://github.com/rob-white/vue-slimgrid#examplevue) is nice)

#### After new Release
- [x] compare file sizes to show in migration 
  - the size is expected to grow since we had to replace certain jQuery functions with our own implementation
  - I checked these 3 files in the root (core, grid & dataview) and the size grew by 18.9Kb but that's unminified
- [x] the [slickgrid.net](http://slickgrid.net/) website should also be updated to drop jQuery & jQueryUI
- [x] look at closing some of the old PRs, and maybe old issues, that are no longer relevant 
- [x] remove jQuery from example4-model, we do it later because this example uses minified dist files

**Comment by ghiscoding** [comment]
@6pac so I think we're ready for a `4.0.0-beta`, just run it on the `next` branch and make sure to use pre-release option, that should be either option 6 or 7. I will then be able to complete the testing in my libs and then write a Migration Guide

![image](https://github.com/6pac/SlickGrid/assets/643976/0f8b94b0-5a97-46f0-ba2c-7bb0f45a0b63)

I also assume that your `.env` file has the correct `GITHUB_TOKEN` now, right?

**Comment by 6pac** [comment]
OK, will make sure I'm not in a hurry or distracted this time ;-)

I did fix up the token, so I think we're good.

**Reply by ghiscoding** [reply]
@6pac alright, I used v4.0 beta from npm and it's all good on my side in my lib, so I went ahead and created PR #777 to merge the `next` branch to `master`, I left the PR open feel free to merge

**Comment by ghiscoding** [comment]
> [ ] the [slickgrid.net](http://slickgrid.net/) website should also be updated to drop jQuery & jQueryUI

@6pac can you look at this one since you are the only one who has access to that 😉 Just mention the jQuery & jQueryUI drop from your website. 

Also note that I added the 2 new Examples you created under the Examples list, I've added them into the [Other Features](https://github.com/6pac/SlickGrid/wiki/Examples#other-features) section... but there seems to be some issues with the new feature, I left a [comment](https://github.com/6pac/SlickGrid/pull/765#issuecomment-1553738518) in PR #765

**Comment by 6pac** [comment]
@ghiscoding working on it but also going to spruce up the site a bit

**Comment by 6pac** [comment]
@ghiscoding done

**Reply by ghiscoding** [reply]
@6pac I'm ready for a final release of v4.0 whenever you are ready, I finished preparing the conversion of all my libs and I'm done testing. You could release without the hidden feature since I'm not using it yet, you could release without it and take your time to fix it later unless you think it's an easy fix. Cheers

**Reply by ghiscoding** [reply]
@6pac not sure if you saw my comment, but I'm done with all my testing and migration so could you proceed with a final v4 release? I'm not using the Hidden Column feature yet, but I probably will in the future, so I don't mind that it's not finalized just yet. Cheers

**Comment by 6pac** [comment]
@ghiscoding yep I did but just very busy this week. Will fix bug and release tomorrow.
